rpcserver.go

```diff
@@ -2936,6 +2939,174 @@ func (s *rpcServer) determineAccountVersion(
+		if transactions[i].BlockHeight != transactions[j].BlockHeight {
+			return transactions[i].BlockHeight < transactions[j].BlockHeight
+		}
+		return transactions[i].TimeStamp < transactions[j].TimeStamp
+	})
+
+	// A map from account key to an ordered list of account action fees.
+	accountActions := make(map[string][]ActionFee)
+
+	for _, tx := range transactions {
+		// Handle account creation action.
+		accountCreation, err := parseAccountCreationAction(tx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse account creation action label: %v", err)
+		}
+		if accountCreation != nil {
+			// Select the channel transaction output amount.
+			outputAmount, err := txOutputAmount(tx)
+			if err != nil {
+				return nil, err;
+			}
+
+			accountActions[accountCreation.AccountKey] = []ActionFee{
+				{
+					"creation",
+					tx.BlockHeight,
+					tx.TimeStamp,
+					*outputAmount,
+					tx.TotalFees,
+				},
+			}
+		}
+
+		// Handle account modification action.
+		accountModification, err := parseAccountModificationAction(tx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse account modification action label: %v", err)
+		}
+		if accountModification != nil {
+			actions := accountActions[accountModification.AccountKey]
+
+			// Calculating the fee from an account modification
+			// action requires the existence of a preceding action.
+			// Skip further processing if preceding action not found.
+			if len(actions) == 0 {
+				continue
+			}
+			precedingAction := actions[len(actions)-1]
+
+			// Select the channel transaction output amount.
+			outputAmount, err := txOutputAmount(tx)
+			if err != nil {
+				return nil, err;
+			}
+
+			// Calculate fee paid for action transaction using the
+			// preceding action.
+			fee := precedingAction.OutputAmount - *outputAmount - tx.Amount
+
+			actions = append(actions, ActionFee{
+				"modification",
+				tx.BlockHeight,
+				tx.TimeStamp,
+				*outputAmount,
+				fee,
+			})
+			accountActions[accountModification.AccountKey] = actions
+		}
+	}
+
+	return accountActions, nil
+}
+
+func txOutputAmount(tx *lnrpc.Transaction) (*int64, error) {
+	for _, outputDetail := range tx.OutputDetails {
+		if !outputDetail.IsOurAddress {
+			return &outputDetail.Amount, nil
+		}
+	}
+	return nil, fmt.Errorf("failed to find output amount, tx hash: %v", tx.BlockHash)
+}
+
+var reAccountCreationLabel = regexp.MustCompile(
+	`poold -- AccountCreation\(acct_key=(?P<acct_key>[0-9a-z]*)\)`,
+)
+
+type accountCreationAction struct {
+	AccountKey string
+}
+
+func parseAccountCreationAction(tx *lnrpc.Transaction) (*accountCreationAction, error) {
+	match := reAccountCreationLabel.FindStringSubmatch(tx.Label)
+	if len(match) > 0 {
+		accountKey := match[1]
+		return &accountCreationAction{accountKey}, nil
+	}
+	return nil, nil
+}
+
+var reAccountModificationLabel = regexp.MustCompile(
+	`poold -- ` +
+	`AccountModification\(` +
+	`acct_key=(?P<acct_key>[0-9a-z]*), ` +
```

Comment:
q: I think this has never changed, but maybe just in case we add/remove more in the future we can do something like:


```
func (s *rpcServer) AccountActionFees() (map[string][]ActionFee, error) {
    ...
    for _, tx := range transactions {
        switch {
        case isAccountCreation(tx):
            processAccountCreation(tx)
        case is AccountModification(tx)
            processAccountModification(tx)
         ... 
         }
         ...
    }
}

func isAccountModification(tx) bool {
    // only match poold --  AccountModification\(`
}

func ProcessAccountModification(tx) {
    // match acc_key
    // match expiry
    // match deposit
    // match is_close
    // separately, so if we change it in the future we are able to parse old tx without the new fields
}

```

Reply:
I've gone with your idea in the latest commit.